### PR TITLE
Fix round off errors in BoxMC with patch potentials.

### PR DESCRIPTION
**Pull request by joaander**
## Description

Prevent round off errors in BoxMC.

## Motivation and context

When the potential energy of the system is large, BoxMC would previously add a small number to a large number, then subtract another large number. This is the potential wash out the effect of the small number (the P * delta V term).

## How has this been tested?

I tested the changes in another repository that checks the physical validity of HOOMD-blue simulations. The change doesn't significantly improve accuracy in my test cases, but it could in others where the potential energy of the system is very large in magnitude.

## Change log

```
* Improve accuracy of ``hoomd.hpmc.update.BoxMC`` when used with patch potentials.
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

### CODE REVIEW

File: hoomd/hpmc/UpdaterBoxMC.cc [hunk 1]
```diff
@@ -259,24 +260,23 @@ inline bool UpdaterBoxMC::box_resize_trial(Scalar Lx,
...
-        // The exponential is a very fast function and we may do better to add pseudo-Hamiltonians
-        // and exponentiate only once...
-        deltaE += ext_energy;
+        delta_U_external = ext_energy;
         }
 
     double p = hoomd::detail::generate_canonical<double>(rng);
 
-    if (allowed && p < fast::exp(-deltaE))
+    if (allowed && p < exp(-delta_beta_H) * exp(-delta_U_pair) * exp(-delta_U_external))
```
Comment (tommy-waltmann) [inline hoomd/hpmc/UpdaterBoxMC.cc:279]: Why remove the `fast::` exponential?

**Comment by joaander** [comment]
FYI, the acceptance criterion for NPT MC is here: http://www.sklogwiki.org/SklogWiki/index.php/Constant-pressure_Monte_Carlo

File: hoomd/hpmc/UpdaterBoxMC.cc (hunk 1 above)
Comment (joaander) [reply hoomd/hpmc/UpdaterBoxMC.cc:279]: This is an accurate exponential computed with doubles. It is computed only once per timestep, so performance is not a concern.

**Review by tommy-waltmann — APPROVED** [review]

**Review by cbkerr — APPROVED** [review]
The changes make sense to calculate the exponents of numbers of different magnitude first.

**Pull request merged**